src/core/lv2/Lv2Ports.cpp

```diff
@@ -0,0 +1,260 @@
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+
+#include "Lv2Ports.h"
+
+#ifdef LMMS_HAVE_LV2
+
+#include "Engine.h"
+#include "Lv2Basics.h"
+#include "Lv2Manager.h"
+
+namespace Lv2Ports {
+
+
+
+
+const char *toStr(Flow pf)
+{
+	switch(pf)
+	{
+		case Flow::Unknown: return "unknown";
+		case Flow::Input: return "input";
+		case Flow::Output: return "output";
+	}
+	return "";
+}
+
+
+
+
+const char *toStr(Type pt)
+{
+	switch(pt)
+	{
+		case Type::Unknown: return "unknown";
+		case Type::Control: return "control";
+		case Type::Audio: return "audio";
+		case Type::Event: return "event";
+		case Type::Cv: return "cv";
+	}
+	return "";
+}
+
+
+
+
+const char *toStr(Vis pv)
+{
+	switch(pv)
+	{
+		case Vis::Toggled: return "toggled";
+		case Vis::Enumeration: return "enumeration";
+		case Vis::Integer: return "integer";
+		case Vis::None: return "none";
+	}
+	return "";
+}
+
+
+
+
+std::vector<PluginIssue> Meta::get(const LilvPlugin *plugin,
+									std::size_t portNum)
+{
+	std::vector<PluginIssue> portIssues;
+	auto issue = [&portIssues](PluginIssueType i, std::string msg = "") {
+		portIssues.emplace_back(i, std::move(msg)); };
+
+	Lv2Manager* man = Engine::getLv2Manager();
+
+	const LilvPort* lilvPort = lilv_plugin_get_port_by_index(
+								plugin, static_cast<uint32_t>(portNum));
+
+	auto portFunc = [&plugin, &lilvPort, &man](
+		bool (*fptr)(const LilvPlugin*, const LilvPort*, const LilvNode*),
+		const char* str) {
+		return fptr(plugin, lilvPort, man->uri(str).get());
+	};
+
+	auto hasProperty = [&portFunc](const char* str) {
+		return portFunc(lilv_port_has_property, str); };
+	auto isA = [&portFunc](const char* str) {
+		return portFunc(lilv_port_is_a, str); };
+
+	std::string portName;
+	{
+		AutoLilvNode nameNode(lilv_port_get_name(plugin, lilvPort));
+		portName = lilv_node_as_string(nameNode.get());
+	}
+
+	m_optional = hasProperty(LV2_CORE__connectionOptional);
+
+	m_vis = hasProperty(LV2_CORE__integer)
+		? Vis::Integer // WARNING: this may still be changed below
+		: hasProperty(LV2_CORE__enumeration)
+		? Vis::Enumeration
+		: hasProperty(LV2_CORE__toggled)
+		? Vis::Toggled
+		: Vis::None;
+
+	if (isA(LV2_CORE__InputPort)) { m_flow = Flow::Input; }
+	else if (isA(LV2_CORE__OutputPort)) { m_flow = Flow::Output; }
+	else {
+		m_flow = Flow::Unknown;
+		issue(unknownPortFlow, portName);
+	}
+
+	m_def = .0f; m_min = .0f; m_max = .0f;
+
+	if (isA(LV2_CORE__ControlPort))
+	{
+		m_type = Type::Control;
+
+		if (m_flow == Flow::Input)
+		{
+			bool isToggle = m_vis == Vis::Toggled;
+
+			LilvNode * defN, * minN = nullptr, * maxN = nullptr;
+			lilv_port_get_range(plugin, lilvPort, &defN,
+					isToggle ? nullptr : &minN,
+					isToggle ? nullptr : &maxN);
```

Comment:
Looking at what `unique_ptr` offers, the only way how I could pass the `unique_ptr` to `lilv_port_get_range` is to use `get()`, but that also won't work (`get` of an unassigned `unique_ptr` will return `nullptr`). So there seems no way around the raw pointers `defN`, `minN` and `maxN`.

I can pack those raw pointers into `unique_ptr` right after the call. That way, the pointers are at least saved right after `lilv_port_get_range`.

Solution without `unique_ptr`:
```
class ParamLilvNodePtr
{
	LilvNode* m_node = nullptr;
public:
	LilvNode* get() { return m_node; }
	LilvNode** nodePtrAddr() { return &m_node; }
	~ParamLilvNodePtr() { lilv_node_free(m_node); }
};
```
It allows to pass `defN.nodePtrAddr()` to `lilv_port_get_range`. Not sure if it really makes anything easier though.

I think there's no easier solution (though we could ask on StackOverflow...).

Reply:
Yeah, I've come across this problem a few times before and never found an elegant solution to it (not for want of a lot of internet searching either). The easiest approach is just to put the raw pointers into `unique_ptr`s straight after the call, as you suggested.

Another approach I've seen is to use a proxy class for putting the pointers into `unique_ptr`s automatically:
```cpp
class PtrRef
{
public:
	explicit PtrRef(AutoLilvNode &node) : m_autoNode(node), m_rawNode(nullptr) { }
	PtrRef(const PtrRef &) = delete;
	PtrRef &operator=(const PtrRef &) = delete;
	~PtrRef() { m_autoNode.reset(m_rawNode); }
	operator LilvNode**() { return &m_rawNode; }

private:
	AutoLilvNode &m_autoNode;
	LilvNode *rawNode;
};

AutoLilvNode defN, minN, maxN;
lilv_port_get_range(plugin, lilvPort, PtrRef(defN), PtrRef(minN), PtrRef(maxN));
```
This is a bit nicer, but at the expense of more code and the potential for some [subtle bugs](https://devblogs.microsoft.com/oldnewthing/20190429-00/?p=102456).

The trouble with a smart pointer which exposes the address of its internal pointer is that there's no way to ensure the old value is freed before it gets overwritten (short of freeing it before returning said pointer, which isn't necessarily desirable). That isn't an issue in this particular case, since you're passing a fresh, unused pointer, but it's not a safe construct in general.

I'm happy for the simple approach to be used here, if you are too - it saves introducing a lot of complication for just one place in the code, and there's not a lot that can go wrong before all the pointers are in place (`unique_ptr`'s constructor is `noexcept`, and I'd like to think anyone maintaining this later wouldn't be silly enough to break control flow between acquiring and storing the pointers).

Reply:
Yeah, I also prefer the simple approach of just using `AutoLilvNode` right afterwards. Looks most comprehensive.